attributable_error_decrypt.go

```diff
@@ -0,0 +1,209 @@
+package sphinx
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+)
+
+// DecryptedAttributableError contains the decrypted attributable error message
+// and its sender.
+type DecryptedAttributableError struct {
+	DecryptedError
+
+	// Payloads is an array of data blocks reported by each node on the (error)
+	// path.
+	Payloads [][]byte
+}
+
+// OnionAttributableErrorDecrypter is a struct that's used to decrypt
+// attributable onion errors in response to failed HTLC routing attempts
+// according to BOLT#4.
+type OnionAttributableErrorDecrypter struct {
+	AttributableErrorStructure
+
+	circuit *Circuit
+}
+
+// NewOnionAttributableErrorDecrypter creates new instance of an attributable
+// error onion decrypter.
+func NewOnionAttributableErrorDecrypter(circuit *Circuit,
+	structure *AttributableErrorStructure) *OnionAttributableErrorDecrypter {
+
+	return &OnionAttributableErrorDecrypter{
+		AttributableErrorStructure: *structure,
+		circuit:                    circuit,
+	}
+}
+
+// DecryptError attempts to decrypt the passed encrypted error response. The
+// onion failure is encrypted in backward manner, starting from the node where
+// error have occurred. As a result, in order to decrypt the error we need get
+// all shared secret and apply decryption in the reverse order. A structure is
+// returned that contains the decrypted error message and information on the
+// sender.
+func (o *OnionAttributableErrorDecrypter) DecryptError(encryptedData []byte) (
+	*DecryptedAttributableError, error) {
+
+	// Ensure the error message length is enough to contain the payloads and
+	// hmacs blocks. Otherwise blame the first hop.
+	if len(encryptedData) < minPaddedOnionErrorLength+o.hmacsAndPayloadsLen() {
```

Comment:
This is good question. 

In master what happens is that a random failure message is made up if the length is below the (legacy) minimum:
https://github.com/lightningnetwork/lnd/blob/e549096b8837a95e0515c8c63fa01c634714d913/htlcswitch/link.go#LL1889C2-L1889C2

For attributable errors though, it may pass this check. Then in the intermediate EncryptError, an error is returned.

I think currently in https://github.com/lightningnetwork/lnd/pull/7139, the intermediate node ignores the error and passes back the failure message as is. This allows an attacker to get away with it.

What should happen instead is that the intermediate node creates a new, properly formatted failure to attribute blame to the connection between itself and its downstream peer.